### DISCUSSION TITLE
TQMesh for 3D Surface Meshing?

### DISCUSSION THREAD

**Discussion by ramcdona** [FloSewn/TQMesh#30] [General]
My program generates isotropic triangle meshes on surfaces in 3D space.

I am interested in considering replacing my algorithms with TQMesh if it can provide some features that my algorithm lacks -- switching may be easier (and better in the long run) than implementing our own.

For example, I currently lack quality quad meshing.

My surfaces are all piecewise Bezier surfaces (potentially high order) -- and so are accessed with U,V parametric coordinates.  The surfaces can have degeneracies (for example, a rectangle in U,V degenerates to a triangle in XYZ).

Would it be possible to modify TQMesh to generate 3D surface meshes by providing it with a function to call that provides a metric tensor (i.e. M(U,V)).  Such that it could operate in two dimensions (U,V), but the size, length, and quality quantities that it uses to progress would be modified by the local metric tensor M?

In addition, does TQMesh support anisotropic meshing?  (another feature I'd like to gain).

**Comment by FloSewn** [comment]
Hey there,

Thanks for your interest - this sounds like a really interesting idea. 
I think your approach might work out - so you perform the mesh generation in U-V-space and then transform it back to 3D. 
Concerning the metric tensor M(U,V), it should be possible to pass it simply as size function, so that the local mesh size is influenced accordingly.

Let me know if I can help you with any details - I am very interested to know if this approach works.

Concerning anisotropic mesh generation: This is currently not supported and I don't have it on my to-do list - but maybe it is worth a try?

Greetings
Flo

**Reply by ramcdona** [reply]
I will probably not be able to take this project on right away, but if its priority rises, I'll be back.  I think the best way to start would be to implement a Metric tensor for a unit sphere in terms of spherical coordinates.  If we can succeed in generating desirable meshes on spheres, it should extend to other curved surfaces.

Often anisotropic mesh generation is also achieved via a metric tensor approach -- you generate isotropic triangles in a stretched space.  It may be possible to get both of these features in a single go.

Thanks for the encouragement.  If I'm able to dedicate time to this, I'll come back and ask for pointers where to begin.
